diff --git a/GNUmakefile b/GNUmakefile
index bb83b14..13f942c 100755
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -96,9 +96,9 @@ endif
 
 # Guard use of -march=native
 ifeq ($(GCC_COMPILER),0)
-   CXXFLAGS += -march=native
+   #CXXFLAGS += -march=native
 else ifneq ($(GCC42_OR_LATER),0)
-   CXXFLAGS += -march=native
+   #CXXFLAGS += -march=native
 else
   # GCC 3.3 and "unknown option -march="
   # GCC 4.1 compiler crash with -march=native.
